File: matlab/QuatMath/Quat2DCMat.m
function C = Quat2DCMat(q)
%Quat2DCMat Converts a quaternion orientation to ZYX Euler angles
%
%   C = Quat2DCMat(q)
%
%   Converts a orientation quaternion (4x1) to corresponding direction 
%   cosine matrix C (3x3) for the 3D rotation
%
    
    CPq = CPMat(q(2:4));
    C = eye(3) - 2*q(1)*CPq + 2*CPq*CPq;

end

File: pycode/QuatMath/Quat2DCMat.py
"""
Quat2DCMat.py

Convert quaternion to direction cosine matrix (rotation matrix)
Original MATLAB sensor fusion algorithms converted for Python/NumPy
"""

import numpy as np


def quat_to_dcm(q):
    """
    Convert quaternion to direction cosine matrix
    
    Args:
        q (array): Quaternion [w, x, y, z]
    
    Returns:
        array: 3x3 rotation matrix
    """
    q = np.array(q).flatten()
    w, x, y, z = q[0], q[1], q[2], q[3]
    
    # Direction cosine matrix
    dcm = np.array([
        [1-2*(y*y+z*z), 2*(x*y-w*z), 2*(x*z+w*y)],
        [2*(x*y+w*z), 1-2*(x*x+z*z), 2*(y*z-w*x)],
        [2*(x*z-w*y), 2*(y*z+w*x), 1-2*(x*x+y*y)]
    ])
    
    return dcm


# Alias for MATLAB compatibility  
Quat2DCMat = quat_to_dcm

